### PR TITLE
Stabilize std::thread

### PR DESCRIPTION
**Pull request by aturon**
This commit takes a first pass at stabilizing `std::thread`:
- It removes the `detach` method in favor of two constructors -- `spawn`
  for detached threads, `scoped` for "scoped" (i.e., must-join)
  threads. This addresses some of the surprise/frustrating debug
  sessions with the previous API, in which `spawn` produced a guard that
  on destruction joined the thread (unless `detach` was called).
  
  The reason to have the division in part is that `Send` will soon not
  imply `'static`, which means that `scoped` thread creation can take a
  closure over _shared stack data_ of the parent thread. On the other
  hand, this means that the parent must not pop the relevant stack
  frames while the child thread is running. The `JoinGuard` is used to
  prevent this from happening by joining on drop (if you have not
  already explicitly `join`ed.) The APIs around `scoped` are
  future-proofed for the `Send` changes by taking an additional lifetime
  parameter. With the current definition of `Send`, this is forced to be
  `'static`, but when `Send` changes these APIs will gain their full
  flexibility immediately.
  
  Threads that are `spawn`ed, on the other hand, are detached from the
  start and do not yield an RAII guard.
  
  The hope is that, by making `scoped` an explicit opt-in with a very
  suggestive name, it will be drastically less likely to be caught by a
  surprising deadlock due to an implicit join at the end of a scope.
- The module itself is marked stable.
- Existing methods other than `spawn` and `scoped` are marked stable.

The migration path is:

``` rust
Thread::spawn(f).detached()
```

becomes

``` rust
Thread::spawn(f)
```

while

``` rust
let res = Thread::spawn(f);
res.join()
```

becomes

``` rust
let res = Thread::scoped(f);
res.join()
```

[breaking-change]

### CODE REVIEW

**Comment by aturon** [comment]
r? @alexcrichton 
cc @pythonesque

**Comment by pythonesque** [comment]
This API makes sense, and it's good to see that this usecase will be supported in the standard library.

The one concern I have is that with this API, you have no choice but to detach the thread immediately, which is a limitation that didn't previously exist.  But I can't immediately think of a reason not to detach immediately if you were going to detach at all, so maybe this is acceptable.

**Comment by aturon** [comment]
@pythonesque We could always re-introduce a `detach` method on `JoinGuard<'static, T>`. I wanted to avoid that for now, to keep our options maximally open.

Thanks for the feedback!

**Comment by kballard** [comment]
I have the same concerns as @pythonesque. I like this API, but I think we should still have a way to explicitly `detach()` when necessary, because I could imagine edge cases where a program can't decide if it wants to detach until the the thread has already started work.

Here's an example of a toy program that requires this ability:

```rust
use std::thread::Thread;
use std::sync::mpsc::channel;
use std::rand::{self, Rng};
use std::io::timer::sleep;
use std::time::duration::Duration;

fn run_thread() {
    let (tx, rx) = channel();
    println!("running thread");
    let guard = Thread::spawn(move || {
        tx.send(rand::thread_rng().gen_range(0f32, 1f32)).unwrap();
        println!("thread sleeping");
        sleep(Duration::seconds(2));
        println!("thread finished sleeping");
    });
    if rx.recv().unwrap() < 0.5 {
        println!("detaching");
        guard.detach()
    } else {
        println!("joining")
    }
}

fn main() {
    run_thread();
    println!("exiting main thread");
}
```

**Comment by kballard** [comment]
> I wanted to avoid that for now, to keep our options maximally open.

I'd suggest reintroducing `detach()` but leave it marked `#[unstable]` or `#[experimental]`, if you don't want to commit to keeping it.

**Comment by aturon** [comment]
Thanks for the feedback @kballard! Reintroducing as `#[experimental]` seems harmless enough; I'll update.

**Comment by aturon** [comment]
Added back `detach`.

**Comment by sfackler** [comment]
Are we okay with panicing if thread creation fails? Should `spawn` and `detach` return `Result`s?

**Comment by pythonesque** [comment]
Ah yeah, that would be useful as well.

File: src/libstd/thread.rs [hunk 1]
```diff
@@ -285,11 +326,13 @@ struct Inner {
...
 
 #[derive(Clone)]
+#[stable]
 /// A handle to a thread.
 pub struct Thread {
     inner: Arc<Inner>,
 }
 
+#[stable]
 unsafe impl Sync for Thread {}
```
Comment (alexcrichton) [inline src/libstd/thread.rs]: This impl shouldn't be necessary, right? The `Arc` implies as much?

File: src/libstd/path/posix.rs [hunk 2]
```diff
@@ -525,17 +525,17 @@ mod tests {
     #[test]
     fn test_null_byte() {
         use thread::Thread;
-        let result = Thread::spawn(move|| {
+        let result = Thread::scopedspawn(move|| {
```
Comment (sfackler) [inline src/libstd/path/posix.rs]: scopedspawn?

**Comment by aturon** [comment]
@sfackler 

> Are we okay with panicing if thread creation fails? Should `spawn` and `detach` return `Result`s?

It's a good question. My and @alexcrichton's instinct here is that this is a relatively niche concern that could be relegated to `std::os::thread` (or something like that, details still being worked out) -- a lower-level, cross-platform thread API.

In any case, these are `#[unstable]` for now so we can revisit during alpha.

@alexcrichton Nits addressed.

File: src/libstd/c_str.rs [hunk 3]
```diff
@@ -0,0 +1,857 @@
+// Copyright 2012 The Rust Project Developers. See the COPYRIGHT
```
Comment (alexcrichton) [inline src/libstd/c_str.rs]: It's back from the dead!

**Comment by thestinger** [comment]
@sfackler: It's not much different from memory allocation failing. It needs to be handled in robust software but Rust's libraries are not usable for that anyway.

File: src/libstd/thread.rs [hunk 4]
```diff
@@ -382,21 +443,24 @@ struct Packet<T>(Arc<UnsafeCell<Option<Result<T>>>>);
...
 unsafe impl<T> Sync for Packet<T> {}
 
-#[must_use]
 /// An RAII-style guard that will block until thread termination when dropped.
 ///
 /// The type `T` is the return type for the thread's main function.
-pub struct JoinGuard<T> {
+#[must_use]
+#[unstable = "may change with specifics of new Send semantics"]
+pub struct JoinGuard<'a, T> {
```
Comment (alexcrichton) [inline src/libstd/thread.rs]: Should this also have `T: 'a` ?

**Comment by alexcrichton** [comment]
This is a pretty conservative step, so r=me with a few nits, we've still got some room to tweak interfaces here and there as necessary.

**Comment by pythonesque** [comment]
@thestinger Failing to initialize a new thread happens a lot more often than failing to allocate memory (at least in my experience).

**Comment by sfackler** [comment]
And conversely, thread creation is comparatively much rarer than memory allocation so the cost of returning a `Result` from thread creation is far lower than returning one from any memory allocation location.

**Comment by pythonesque** [comment]
On the implementation side, would it be better to use Option<Thread> instead of having a "joined" flag?  That should save a word in the struct size.

**Comment by aturon** [comment]
@pythonesque @sfackler @alexcrichton I talked with some of you all about this on IRC, but: I think a good middle ground here is for the `spawn` and `scoped` methods on the builder to return `Result`s, but on `Thread` (where they are meant as conveniences) to `panic` in these situations.

Usually we avoid such duplication (i.e., no `spawn` and `spawn_result` variants), but here the API is already duplicated for convenience; the builder is a kind of power tool. This seems to strike a good balance between making the common case easy and providing the tools for extra robustness if you want them.

All that said, I'd like to move forward with this PR as-is since there's a limited window to land these changes before alpha. Since the relevant bits of the API are `#[unstable]`, we can revisit shortly after the alpha release.

**Comment by pythonesque** [comment]
Seems reasonable to me.

**Comment by sfackler** [comment]
:+1:
